### ISSUE TITLE
The case for Thread.Abort / Thread.Suspend

### ISSUE THREAD

**Issue by pongba** [dotnet/coreclr#20705]
So Thread.Abort / Thread.Suspend are not just deprecated, they're removed outright from .NET Core, which I think is unnecessary.

.NET Core's nature is to power performance-critical online services, and for services, being "alive" is everything. A lot of efforts can be spent on making the service resilient against unexpected events.

.NET Core is also a very good embedding engine to power a plug-in like system (the service stays native, and extension code runs in .NET Core embedded in the host native service), it's not always realistic to expect that all code running on top of .NET Core to cooperatively check for cancellation tokens, nor would that completely eliminate the risk of "3rd party" extension code misbehaving.

Therefore, as a last resort, it's often required to do whatever is in the power of the service to stay live and being able to serve requests, **even at the risk of having partially corrupted state and/or partially inconsistent state**, for two reasons:

1. for a well-designed plug-in system, it can be designed so that any state the plug-in code can access & manipulate is per-request, so corruption to the state doesn't have global effects.
2. not every service is so sensitive to state corruption like database services, often times, arguments can be made that partially correct results can be better than no results. 

That is, even if Thread.Abort can lead to inconsistent or corrupted state, it shouldn't be the reason to remove it all together from our toolbox. 

We should trust service engineers' ability to weigh the risk against the benefit to make decision on their particular case, instead of worrying that they will shoot their foot somehow.

I think the extreme insistence on a "no corrupted state" program (hence the mindset "just crash if state is corrupted") doesn't apply well to today's service oriented world.

Bottom line is, without these two APIs, I have *no way* to stop suspected infinitely looping threads from eating up my cores and bring down my service's capacity, which is a big deal for services.

**Comment by GSPP** [comment]
I think the case you make is very reasonable. It is also my experience that the risk trade-off for aborts can be worth it. ASP.NET for example implements request timeouts with `Thread.Abort`. But this approach can absolutely destroy the entire app domain (for example by crashing a static constructor). I usually disable the request timeout entirely therefore.

I personally have never seen a good case for `Thread.Abort` but I'm sure it exists. You seem to be in a place where you sometimes need to abort 3rd party code as a mitigation of last resort. It certainly is a radical stance to remove `Thread.Abort` from the platform and prevent these good use cases.

It would be incorrect to say that the corruption risk of `Thread.Abort` is *never* worth it. Real world applications sometimes require nasty trade-offs.

I'm personally very happy to see the scourge that is `Thread.Abort` to be purged from the platform. So much evil has been committed using it. It's a net benefit to prevent it from being used even considering that valid use cases are prevented as well.

**Comment by pongba** [comment]
@GSPP thanks for the detailed and considerate response. I have actually encoutered a case where I would like to be able to abort a thread - the thread executes some extension code not owned by the service owner, if it end up busy looping, it will eat up my cores, bring down service capacity. 

To restart service would be a too aggresive move, and probably won't help anyways, since new request can still make the situation re-occur.

Only viable damage control is to stop further execution of such misbehaving code - but what's already damaged can not be reversed, i.e. we can stop more cores from being further eaten up, but what's already eaten up (we of course set a maximum limit N) cannot be brought back unless we can suspend or abort the bad thread.

P.S. I don't think the static constructor situation is always a big concern - sure it can bring down the whole appdomain, but first of all, this is only a disaster for managed service, since the whole service may relie on the appdomain to function, but for hybrid services (becoming more & more common nowadays), the dying of an appdomain is simply one fault that can be tolerated or even partially recovered from (say recreate the appdomain, or recreate the .NET Core CLR), second of all, isn't there ways to make sure a certain area of code (e.g. extension code) never uses static constructors through roslyn code analysis?

**Comment by benaadams** [comment]
>  but for hybrid services (becoming more & more common nowadays), the dying of an appdomain is simply one fault that can be tolerated

Core CLR doesn't have appdomains; it will crash the process

**Comment by pongba** [comment]
@benaadams I know. But not if CoreCLR is hosted by a native process.

**Comment by benaadams** [comment]
Could use multi-process and terminate the process instead? [Porting to .NET Core - AppDomains](https://docs.microsoft.com/en-us/dotnet/core/porting/libraries#appdomains)

> For code isolation, we recommend separate processes or using containers as an alternative.

**Comment by pongba** [comment]
> ASP.NET for example implements request timeouts with `Thread.Abort`. But this approach can absolutely destroy the entire app domain (for example by crashing a static constructor). I usually disable the request timeout entirely therefore.

I completely agree that Thread.Abort should not be used as a timeout mechanism, because timeout is a very frequent event, and repeatedly Thread.Aborting is not a pretty scene to imagine. 

In my mind, Thread.Abort is _never_ supposed to be used for timeout.

It's instead used for "this-thread-is-definitely-doing-evil-stuff-like-infinitely-looping-or-blocking", which is a whole different category than normal everyday timeout. Admittedly it's theoretically impossible to determine if a thread is indeed in this kind of state, but in practice it's often possible to do so through domain knowledge particular to one's particular service and application; a very simple way that often works is to have a very relaxed timer that should the timer fires, then definitely things are going south. And this timer is of course conceptually different from the ordinary timeout timer (usually it's much longer)

**Comment by pongba** [comment]
> Could use multi-process and terminate the process instead? [Porting to .NET Core - AppDomains](https://docs.microsoft.com/en-us/dotnet/core/porting/libraries#appdomains)
> 
> > For code isolation, we recommend separate processes or using containers as an alternative.

Yes. That can work. But that kind of destroys the whole essence of .NET Core's main battleground - extreme performance. Not to mention that it's often a huge a mount of dev work to make the process level isolation work nicely if the communication protocol between the processes are not "trivial" such as a blob of data tossed over (shared memory of course) the boarder and then response blob of data received once the other process has done its job.

**Comment by GSPP** [comment]
> isn't there ways to make sure a certain area of code (e.g. extension code) never uses static constructors through roslyn code analysis?

Even `System.String` has a static constructor. All kinds of types initialize caches using them. That said all of these would typically occur in the first few moments of an application's execution. It does not seem to be a huge problem in practice but it bears the potential of permanently bring down the appdomain.

**Comment by mnmr** [comment]
The documentation for Thread.About is very specific about what the method can and cannot do: "It might terminate the thread". That's not a great API to support in .NET Core.

Unless Windows starts offering something with better guarantees I fully support the decision to NOT have these mechanisms in .NET Core. There is also the concern that whatever mechanism .NET Core introduces, it should work across operating systems.

You can start multiple processes and use that to provide reliability. It's not as convenient as handling it internally in a single process, but you can do it without using API's that are documented "to maybe work".

**Comment by Suchiman** [comment]
@mnmr this has little to do with windows offering something or cross platform. Calling `Abort` throws an asynchronous exception (async because it is not caused by something that thread is doing itself) into the target thread. That can only happen in managed code when not in finally blocks and will only abort if the target thread doesn't cancel the abort by catching it and calling `ResetAbort`. Those are well defined behaviors. Similiar one could argue that starting another process is not a great API to support in .NET Core either because it "only maybe works" since you can run out of system resources necessary to start a new process.

**Comment by pongba** [comment]
@mnmr Multi-process is often a costly option that goes against the the latency requirement. 

And whether an API has guaranteed success or not shouldn't be the reason for banning it outright. As @Suchiman said, they're orthogonal matters. We have a whole lot of APIs that "may or may not" succeed.

I would argue that it's a philosophical question eventually, i.e. whether to trust engineers' ability to make judgments on their own, or babysit them by taking away the sharp objects from the playpen. 

In the past where .NET's major battleground is not extreme performance-sensative areas, I would say the babysitting approach is fine, but, nowadays we see .NET Core extending deep into so many extremely performant areas where not only it's required to have riskier APIs, but also where service engineers working there are capable of making risk-aware choices for themselves.

The bottom line is: I don't really see the downside of including the API, since with the current state of C# development environment, we can state very clearly the danger (when and how, and the alternatives, if they apply) in "in-code documentations" such as [Obsolete] tag, and enforce them through warnings, code analysis suggestions, and etc. With these on-the-spot educational tips while the code is being written, I really don't see the over worrying sentiment behind this API, after all they have been included in Windows API like forever.

**Comment by Suchiman** [comment]
> I would argue that it's a philosophical question eventually, i.e. whether to trust engineers' ability to make judgments on their own, or babysit them by taking away the sharp objects from the playpen.

This is something particular observed well in .NET Core, prior examples include removing `StackTrace` because some developers abused it in an ORM to check for transaction nesting.

**Comment by pongba** [comment]
@Suchiman yup, and I worry about the philosophy behind these kind of decision makings, **particularly** because this is .NET Core, this is in and of itself the sharp sword we're handing over to service engineers who would go to great length to squeeze out the last drop of performance gain, and now we're worried about them cutting themselves because we didn't remove some rough edges on the sword's handle. This is paradoxical. Why then don't we ban "unsafe" all together?

And trust me I want .NET Core successful as much as anyone.

And I even want a Span<T> that directly holds pointers to native heap (to reduce interop cost to essentially 0; as of now to my understanding of Span<T> if I need to expose a native std::string's guts to managed code, 1 copy of memory is still needed.)

**Comment by svick** [comment]
@pongba 

> I worry about the philosophy behind these kind of decision makings, particularly because this is .NET Core, this is in and of itself the sharp sword we're handing over to service engineers who would go to great length to squeeze out the last drop of performance gain, and now we're worried about them cutting themselves because we didn't remove some rough edges on the sword's handle. This is paradoxical. Why then don't we ban "unsafe" all together?

Experienced developers who care a lot about performance are not the only target audience of .Net Core. It's also meant to be used by beginners, and everyone in between the two extremes, so I don't see the paradox. Performance is important, but it's not the only important thing.

And the difference between `unsafe` and `Thread.Abort` is that `unsafe` is hard to use safely, while it's generally considered impossible to use `Thread.Abort` safely.

> And I even want a Span that directly holds pointers to native heap (to reduce interop cost to essentially 0; as of now to my understanding of Span if I need to expose a native std::string's guts to managed code, 1 copy of memory is still needed.)

[`Span<T>` has a constructor that takes a pointer an length](https://docs.microsoft.com/en-us/dotnet/api/system.span-1.-ctor?view=netcore-2.1#System_Span_1__ctor_System_Void__System_Int32_), so I don't see why accessing `std::string::data` though `Span<byte>` would require a copy. Though I think that discussion is off-topic on this issue.

**Comment by pongba** [comment]
> [`Span<T>` has a constructor that takes a pointer an length](https://docs.microsoft.com/en-us/dotnet/api/system.span-1.-ctor?view=netcore-2.1#System_Span_1__ctor_System_Void__System_Int32_), so I don't see why accessing `std::string::data` though `Span<byte>` would require a copy. Though I think that discussion is off-topic on this issue.

@svick Thanks a lot, it appears that my readings were off. This is great to know.

**Comment by pongba** [comment]
> And the difference between `unsafe` and `Thread.Abort` is that `unsafe` is hard to use safely, while it's generally considered impossible to use `Thread.Abort` safely.

@svick then put a huge warning sign in the API that basically says "you think you get this, but you don't. read the full document and you'll see why". stuff like that.

I also noted that you agree that it's not impossible to use correctly (or more precisely, used in the correct scenario). When it's used in the correct scenario it can be very useful, and removing a very useful tool out of the toolbox is not a good design, especially considering the fact that modern development environment has so many ways to discourage or forbid certain APIs in a user-specific way (it should be very easy to have a project-global supression of certain dangerous APIs if project leads consider these taboo), this may even be able to configured in default CSharp build targets config so that any use of it results in error (but can be specifically turned on in a global settings)

**Comment by svick** [comment]
@pongba 

> I also noted that you agree that it's not impossible to use correctly (or more precisely, used in the correct scenario).

I think that either I misunderstand what you're saying or you misunderstood what I said. I do think it is impossible. In your original post, you said:

> for a well-designed plug-in system, it can be designed so that any state the plug-in code can access & manipulate is per-request, so corruption to the state doesn't have global effects.

How exactly do you prevent a plug-in from using global state like Reflection?

If you don't prevent it, then there is no such thing as "partially corrupted state", the whole appdomain is corrupted.

If you do prevent it, then you're almost certainly using something like Roslyn or Cecil to analyze the plug-in code. And if you're doing that, couldn't you also ensure that the plug-in either correctly reacts to cooperative cancellation, or even rewrite it to react appropriately? (For inspiration, [here is the code used by SharpLab for this purpose](https://github.com/ashmind/Unbreakable/blob/693250f/%5BCore%5D/AssemblyGuard.cs).)

**Comment by juliusfriedman** [comment]
@jkotas, perhaps outlining the differences in BeginCritialRegion or something else would be helpful for comparisons sake to readers who are not familiar with the technical semantics associated...

None the less; I feel this feature could have been provided with enough care and also would have allowed other conversations e.g. how one would be able to customize and address compatibility.

Things aside, there's nothing stopping this being stubbed into most projects via an extension method and those who can't due to complexity or other concerns must have been doing it wrong.

**Comment by pongba** [comment]
@jkotas I absolutely agree with you that it's virtually impossible to write normal code that is robust against asynchronous exceptions such as ThreadAbort. However, my main point is that *this* should not be the reason to remove the API from the toolbox, because:

1) It's very dangerous, for sure. And it's extremely easy to misuse (engineers are prone to mistakenly assume Thread.Abort's safetiness). But this danger can (or at least should) be easily controlled by project or even repo-wise API restriction policy set by project leads.

2) In the circumstances where the services *don't* need normal code to be resilient (i.e. when it does Abort it has a very good idea of where the thread is running at), the Abort / Suspend API can save precious cores by cancelling busy-yet-doing-nothing threads. 

A good example is this: at least in two circumstances I have heard this issue of race-condition in concurrent code causing managed Dictionary's internal linked list to form a loop and therefore causing infinite loops that eats up service machine's cores one by one, eventually leading to big service healthiness issues. Note that currently we have *no* way of auto-mitigating this issue once it happens in an online system, unless Thread.Abort is supported.

So to sum up my argument: 1) the downside can be quite easily controlled on a per-project(repo) basis 2) the upside is vital **and irreplacable* to some *important* scenarios.

Last but not least, speaking from the point of view of a 1st party, as online service engineer, who's trying to put together a service that allows extension code from 3rd party, the fact that I cannot have 100% guaranteed way of stopping 3rd party code wakes me up at night, and I would imagine a lot of other engineers would share the same sentiment. (also note that, as a comparison, Javascript V8 engine offers ganranteed termination through v8::Isolate::TerminateExecution API, which offers platform abosolute control over extension code in the form of .js)

All in all, what would you do in such scenario?

**Comment by jkotas** [comment]
> I have heard this issue of race-condition in concurrent code causing managed Dictionary's internal linked list to form a loop and therefore causing infinite loops that eats up service machine's cores one by one

This specific issue has been fixed in .NET Core. The Dictionary in .NET Core will throw exception when it detects corrupted state due to race condition instead of ending up infinite loop. If there are other bugs that you need `Thread.Abort` to workaround, please do let us know. We would rather fix them.

> Javascript V8 engine offers ganranteed termination

The JavaScript language and libraries are designed to allow in-process sandboxing - running arbitrary 3rd party code and abort it reliably without killing the process. We have given up on sandboxing with .NET a long time ago.

**Comment by jkotas** [comment]
> why .NET sandboxing effort has been given up

It is hard to tell where it is "safe" to insert the thread aborts. Libraries that need to be robust in presence of Thread.Abort need to be annotated for it, coded in a special way and stress tested. 

We have tried to do this in .NET Framework: It was a full time job for several people to run a stress harness that inserted thread about at random points in .NET Framework, and file and fix bugs on the crashes, hangs and data corruptions that it hit. This was done only for a subset of .NET Framework that was usable in SQLCLR, and still it was never ending stream of issues.

Even with this effort, we often got a support escalation (from paid support) where people hit problems with Thread.Abort in production. Some of these issues require a very ugly hacks to workaround because of there was just no right fix to them. We had to resort to hacks like decoding assembly instructions and  suppress or adjust thread abort behavior for particular instruction pattern that was known to hit the problem.

I am not even talking about larger .NET ecosystem - if you take a random NuGet package from nuget.org, it is almost guaranteed that it has reliability bugs in the presence of thread abort.

**Comment by benaadams** [comment]
>  good to know it's fixed in .NET Core. I still wonder if the fix adds perf penalties or not.

It doesn't as it https://github.com/dotnet/coreclr/pull/16991 piggybacks on the existing Hash DoS protection; also you will find the Dictionary in .NET Core has had many performance improvements made to it and is much faster than .NET Framework even with this additional protection.

**Comment by jkotas** [comment]
> data structures designed for single-threaded usage shouldn't get worried about its internal invariants getting broken because its public methods are called in a non-thread-safe way

I agree with this in general. However, if we find that a particular datastructure tends to get used incorrectly, it is worth it to make this incorrect use diagnose if it is cheap to do.

**Comment by pongba** [comment]
> > why .NET sandboxing effort has been given up
> 
> It is hard to tell where it is "safe" to insert the thread aborts. Libraries that need to be robust in presence of Thread.Abort need to be annotated for it, coded in a special way and stress tested.
> 
> We have tried to do this in .NET Framework: It was a full time job for several people to run a stress harness that inserted thread about at random points in .NET Framework, and file and fix bugs on the crashes, hangs and data corruptions that it hit. This was done only for a subset of .NET Framework that was usable in SQLCLR, and still it was never ending stream of issues.
> 

@jkotas I really appreciate the detailed explanation, now I see the scope of effort needed to support the API much clearly. And it also appears to me that as much as I would like to wish the API restriction to be an effective way of limiting misuse, it's probably not going to be so easy in real world, people will still somehow find ways or simply ignore the warning sign.

**Comment by AgentFire** [comment]
Core printing API has some serious issues (including outright hangs on a PrintVisual/PrintDocument calls), can else should I be able to stop the freezed thread?

**Comment by Leonardo-Ferreira** [comment]
Well, very good points all over the place here! here are my 2 cents:
 
-A insecure, error prone, "deadly" api is better than no api at all. 
Right now I'm looking at a rancher monitor that shows me ever-growing CPU consuming pods. When the CPU baseline/idle is around 50%, we kill the pods and let Kubernetes recreate them. "Oh the thread aborting corrupted the app domain and crashed the app?" well, it only anticipated the inevitable and now I don't have to keep 50% of my pod doing trash or w8ing for answers that will never come.
Kestrel is a very good and performant web server, but it's inability to timeout a request (wether by aborting a thread, or killing a task) makes it a ticking bomb in hands of lower-tier developers...
